### PR TITLE
Add VM schedules pages

### PR DESCRIPTION
**Pull request by a110605**
### Summary
- add VM schedule list / edit / details pages
- add VM schedule column in vmbackup / vmsnapshot list pages
- add `Create VM Schedule`action for VM operation menu
- show `volume backup error message` and `readyToUse` in backup/snapshot detail volume tab <- discussion conclusion with the team
- upgrade `cron-validator` and `cronstrue` to the latest version

Note. this UI PR needs wait for backend PR https://github.com/harvester/harvester/pull/6149 merge first.
#### PR Checklist
- Is this a multi-tenancy feature/bug?
    - [ ] Yes, the relevant RBAC changes are at:
- Do we need to backport changes to the [old Rancher UI](https://github.com/rancher/u), such as RKE1?
    - [ ] Yes, the relevant PR is at:
- Are backend engineers aware of UI changes?
    - [x] Yes, the backend owner is: @WebberHuang1118 

Related Issue #
https://github.com/harvester/harvester/issues/6058

### Technical notes summary
run `yarn install` before run up the harvester dashboard

### Screenshot/Video

1. Create VM Schedule

https://github.com/user-attachments/assets/45bb297e-b184-4054-a0db-6d5521dac87d

2. Edit VM Schedule

https://github.com/user-attachments/assets/3a7ffdda-4138-4486-9eb5-82265830b2a4

3. See VM Schedule detail

https://github.com/user-attachments/assets/0d525371-2048-413e-86bd-7ff3361803ad

4. show error banner if not set backup target

https://github.com/user-attachments/assets/ba2bcadd-e65a-43db-bc64-3be6f286c17d

5. When create back up failed and reach Max Failure threshold, the schedule job would auto suspended.

auto_stop_schedule

6. Show `volume backup error message` and `readyToUse` in backup/snapshot detail volume tab

https://github.com/user-attachments/assets/f3c42dbf-5d7d-4d63-a1f2-162c5fd9e301

### CODE REVIEW

**Comment by a110605** [comment]
Let me rebase and test it again since backend PR has some minor change.

**Comment by a110605** [comment]
Hi @WebberHuang1118 , @torchiaf , I've rebase this PR and ready to review.

File: pkg/rancher-components/package.json [hunk 1]
```diff
@@ -32,8 +32,8 @@
     "@vue/test-utils": "1.2.1",
     "babel-eslint": "10.1.0",
     "core-js": "3.25.3",
-    "cron-validator": "1.2.0",
```
Comment (torchiaf) [inline pkg/rancher-components/package.json:35]: I think we don't need to update the chron library for `rancher-components`, but It doesn't hurt.
The new lib version is something we have to test when doing the migration though.

**Review by torchiaf — CHANGES_REQUESTED** [review]
I'm working on the review, before I look at the code, I found the  following error, could you check ?

- Virtual Machine Snapshots title starts with ':'
- `Cannot read properties of undefined (reading 'linkFor')` when click create button.

**Comment by a110605** [comment]
> I'm working on the review, before I look at the code, I found the following error, could you check ?
> 
> * 'Create Schedule' title starts with ':'
> * `Cannot read properties of undefined (reading 'linkFor')` when click create button.
> 
> image

The [backend PR](https://github.com/harvester/harvester/pull/6149) is merged. Could you use master-head deployment and try again?

Should be looks like
Screenshot 2024-09-23 at 11 12 39 AM

**Comment by a110605** [comment]
> This pull request is now in conflict. Could you fix it @a110605? 🙏

Done rebase.

**Comment by a110605** [comment]
> LGTM, thans for the PR.
> 
> Just a point, if the schedule is suspended, the GUI doesn't show the condition?
> 
> For example, if the user suspends the schedule proactively, `ScheduleVMBackup` CR's `status.conditions` will have related message and reason.

I have `stateDescription()` grabbing from `this.state.conditions.message`, will look like

Screenshot 2024-09-24 at 4 41 24 PM

Screenshot 2024-09-24 at 4 38 14 PM

**Review by WebberHuang1118 — APPROVED** [review]

**Review by bk201 — APPROVED** [review]
The UX lgtm. However, I feel the word "VM Schedule" is confusing. Maybe "VM Backup Schedule" is more clear?
@jillian-maroket @asettle What do you think about this term.

**Comment by a110605** [comment]
> The UX lgtm. However, I feel the word "VM Schedule" is confusing. Maybe "VM Backup Schedule" is more clear? @jillian-maroket @asettle What do you think about this term.

But user can still create snapshot schedule job from this page, that why I give it `Virtual Machine Schedule`.

File: pkg/harvester/components/FilterVMSchedule.vue [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+<template>
+  <div class="vm-schedule-filter">
+    <template>
+      <span v-if="selected" class="banner-item bg-warning">
+        {{ t('harvester.tableHeaders.vmSchedule') }}{{ selected ? ` = ${selected}`: '' }}<i class="icon icon-close ml-5" @click="remove" />
+      </span>
+    </template>
+
+    <v-popover
+      trigger="click"
```
Comment (torchiaf) [inline pkg/harvester/components/FilterVMSchedule.vue:72]: There is a small empty box showing up when clicking the filter, but there are not options to show:

image

To fix it:

```suggestion
     :trigger="scheduleOptions.length ? 'click' : 'manual'"
```

File: pkg/harvester/detail/harvesterhci.io.schedulevmbackup/BackupList.vue [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+
+    hasBackupProgresses() {
+      return !!this.rows.find(R => R.status?.progress !== undefined);
+    },
+
+    filteredRows() {
+      let r = this.rows.filter(row => row.spec?.type === BACKUP_TYPE.BACKUP);
+
+      if (this.id) {
+        r = r.filter(backup => backup.metadata.annotations?.['harvesterhci.io/svmbackupId'] === this.id);
```
Comment (torchiaf) [inline pkg/harvester/detail/harvesterhci.io.schedulevmbackup/BackupList.vue:93]: We should add 'harvesterhci.io/svmbackupId' in `label-annotation.js`

File: pkg/rancher-components/src/components/Form/LabeledInput/LabeledInput.vue [hunk 4]
```diff
@@ -144,11 +144,16 @@ export default (
       if (this.type !== 'cron' || !this.value) {
         return;
       }
-      if (!isValidCron(this.value)) {
```
Comment (torchiaf) [inline pkg/rancher-components/src/components/Form/LabeledInput/LabeledInput.vue:147]: We should port this fix in rancher/dashboard.

**Review by torchiaf — CHANGES_REQUESTED** [review]
I left some comments and there are some small regressions:

- When you click on 'Replace existing' / 'Restore New' from Virtual Machine Snapshots page or Virtual Machine Schedules -> Snapshots tab:

  1. The UI is redirected to Backups create page rather than Snapshots create.
  2. If you hit Cancel the UI is redirected to Backups page. The redirection should be to Virtual Machine Snapshots or Virtual Machine Schedules -> Snapshots tab, depending on navigation history.

Virtual Machine Snapshots

[Screencast from 2024-09-26 14-37-07.webm](https://github.com/user-attachments/assets/fcd4da4a-5a32-4ae3-96ef-ad50bb9dc2e6)

Virtual Machine Schedules -> Snapshots tab

Consider that some button actions are in common between Snapshots and Backup models.

File: pkg/harvester/models/harvesterhci.io.virtualmachinebackup.js [hunk 5]
```diff
@@ -81,23 +80,22 @@ export default class HciVmBackup extends HarvesterResource {
   }
 
   restoreExistingVM(resource = this) {
-    const route = this.currentRoute();
     const router = this.currentRouter();
 
     router.push({
       name:   `${ HARVESTER_PRODUCT }-c-cluster-resource-create`,
-      params: { resource: route.params.resource },
+      params: { resource: HCI.BACKUP },
```
Comment (torchiaf) [inline pkg/harvester/models/harvesterhci.io.virtualmachinebackup.js:87]: I think this is common with Snapshots functionalities, so I guess this is causing the wrong route error when clicking restore in Snapshots page.

File: pkg/rancher-components/src/components/Form/LabeledInput/LabeledInput.vue (hunk 4 above)
Comment (a110605) [reply pkg/rancher-components/src/components/Form/LabeledInput/LabeledInput.vue:147]: will have PR to add in rancher/dashboard.

**Comment by a110605** [comment]
@WebberHuang1118 , could you help check why I can't `restore new` / `restore existing` snapshot or backup using schedule job created backup ?

image

File: pkg/harvester/models/harvesterhci.io.virtualmachinebackup.js [hunk 6]
```diff
@@ -81,23 +80,22 @@ export default class HciVmBackup extends HarvesterResource {
...
 
   restoreNewVM(resource = this) {
-    const route = this.currentRoute();
+    // const route = this.currentRoute();
     const router = this.currentRouter();
 
     router.push({
       name:   `${ HARVESTER_PRODUCT }-c-cluster-resource-create`,
-      params: { resource: route.params.resource },
+      params: { resource: HCI.BACKUP },
```
[the remark this answers is not in the document]
Comment (a110605) [reply pkg/harvester/models/harvesterhci.io.virtualmachinebackup.js]: fixed in eabd9435

File: pkg/harvester/models/harvesterhci.io.virtualmachinebackup.js (hunk 5 above)
Comment (a110605) [reply pkg/harvester/models/harvesterhci.io.virtualmachinebackup.js]: fix in eabd9435

**Review by torchiaf — APPROVED** [review]
LGTM UI side.

**Comment by torchiaf** [comment]
> @WebberHuang1118 , could you help check why I can't `restore new` / `restore existing` snapshot or backup using schedule job generated backup / snapshot? It's fine to restore snapshot/backup which is manually created.
> 
> image

IMHO the 500 error should be handled backend side, but the error message is consistent: If I stop the VM schedule from which the VM backups is created, I can manually restore the backup into a new VM.

**Comment by WebberHuang1118** [comment]
> @WebberHuang1118 , could you help check why I can't `restore new` / `restore existing` snapshot or backup using schedule job generated backup / snapshot? It's fine to restore snapshot/backup which is manually created.
> 
> image

Yes, the Harvester controller has the webhook to [prevent restoring the snapshot/backup from a running schedule](https://github.com/harvester/harvester/blob/606f485579d087def560e2a8a255a2b0f4bad019/pkg/webhook/resources/virtualmachinerestore/validator.go#L103-L106), since the backup/snapshots will be cleansed under the hood, leading to restore failure. By design, before restoring, the user should suspend the schedule first.

**Pull request merged**